### ISSUE TITLE
Is there any way to interrupt a transfer cleanly?

### ISSUE THREAD

**Issue by TrevorPT** [mscdex/ssh2-streams#29]
Hi, is there any way to stop a transfer in progress? I'm using sftp.fastPut with a progress dialog that contains a "stop" button. I can't find a way to cleanly terminate the put other than using `connection.end()` but I'm not sure if this is a clean termination.  I was considering looking at putting a stop mechanism in `onstep` so it would at least close the handles properly and use the `onerror` callback.  Is there a better way that I'm missing?

Thanks.

**Comment by mscdex** [comment]
If you need to support stopping, you could just copy the code for `fastPut()` and tweak it to your liking. Or you could use `sftp.createWriteStream()`, but it is slower because it only does one write at a time.

**Comment by TrevorPT** [comment]
Thanks for the tip about using `createWriteStream`; I'll have to try it out versus `fastPut`.  For some reason I kept thinking 'concurrency' meant multiple files, not multiple read/writes on one file, so missed the speed benefit there.  It would also likely make stopping the multiple concurrent reads a bit trickier.

**Comment by mscdex** [comment]
Stopping reads is easy, just have some boolean variable that you check before performing a read. If it's false, stop (and possibly remove partial file on remote end).

**Comment by TrevorPT** [comment]
Ah, right.  I suppose each existing write will end and then stop on the next read.  Do you see any issues with just forcibly closing the connection?  Will it likely leak resources (other than leaving a file fragment on the FTP server)?

**Comment by mscdex** [comment]
There shouldn't be any resources leaked if you end the connection early (or you might try just ending the sftp session instead: `sftp.end()`).

**Comment by TrevorPT** [comment]
I modified fastXfer to handle stopping properly on all read threads, though it waits until the writes are complete.  I may play with it some more to refine it further or look at using `.end()` which did work in a quick test though I didn't examine resource consumption for leaks.  I also compared fastPut using a concurrency of 1 versus 25 (I figured it would be similar in performance to using `createWriteStream`) and saw the throughput advantage of the concurrent writes.  Nice!

Thanks for the feedback.
